grammar.js

```diff
@@ -1,284 +1,391 @@
-function commaSep1(rule) {
-  return seq(rule, repeat(seq(',', rule)))
+/* eslint-disable camelcase */
+/* eslint-disable-next-line spaced-comment */
+/// <reference types="tree-sitter-cli/dsl" />
+// @ts-check
+
+/**
+ * Creates a rule to match one or more of the rules separated by the separator
+ * and optionally adds a trailing separator (default is false).
+ *
+ * @param {Rule} rule
+ * @param {string} separator - The separator to use.
+ * @param {boolean?} trailing_separator - The trailing separator to use.
+ *
+ * @return {SeqRule}
+ *
+ */
+const listSeq = (rule, separator, trailing_separator = false) =>
+  trailing_separator ?
+    seq(rule, repeat(seq(separator, rule)), optional(separator)) :
+    seq(rule, repeat(seq(separator, rule)));
+
+/**
+ * Creates a rule to match one or more of the rules separated by a comma
+ * and optionally adds a trailing separator (default is false).
+ *
+ * @param {Rule} rule
+ * @param {boolean?} trailing_separator - The trailing separator to use.
+ *
+ * @return {SeqRule}
+ *
+ */
+function commaSep1(rule, trailing_separator = false) {
+  // return seq(rule, repeat(seq(',', rule)))
+  return listSeq(rule, ',', trailing_separator);
 }
 
-function commaSep(rule) {
-  return optional(commaSep1(rule))
+/**
+ * Creates a rule to optionally match one or more of the rules separated
+ * by a comma and optionally adds a trailing separator (default is false).
+ *
+ * @param {Rule} rule
+ * @param {boolean?} trailing_separator - The trailing separator to use.
+ *
+ * @return {ChoiceRule}
+ *
+ */
+function commaSep(rule, trailing_separator = false) {
+  return optional(commaSep1(rule, trailing_separator));
 }
 
+/**
+ * This callback should take a rule and places it in between a group (e.g. parentheses).
+ * @callback GroupingCallback
+ * @param {Rule} rule
+ *
+ * @return {SeqRule}
+ */
+
+/**
+ * Creates a rule that matches a binary operator in between two rules.
```

Comment:
This looks like it was copy-pasted. It should read: "Creates a rule that accepts another rule with optional grouping symbols".